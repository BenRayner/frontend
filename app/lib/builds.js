export function shortMessage(message) {
  return message.split("\n")[0];
}

export function shortCommit(commitish) {
  // Does this look like a git sha?
  if (commitish && commitish.length === 40 && !commitish.match(/[^a-f0-9]/i)) {
    return commitish.substring(0, 7);
  } else {
    return commitish;
  }
}

export function buildTime(build) {
  const { state, startedAt, canceledAt, finishedAt, scheduledAt } = build;
  const buildTime = {};

  switch (state) {
    case 'failed':
    case 'passed':
    case 'canceled':
    case 'canceling':
    case 'started':
      if (state === 'canceled' && !startedAt) {
        break;
      }

      buildTime.from = startedAt;
      break;

    case 'scheduled':
      buildTime.from = scheduledAt;
      break;
  }

  switch (state) {
    case 'failed':
    case 'passed':
      buildTime.to = finishedAt;
      break;

    case 'canceled':
      if (!startedAt) {
        break;
      }

      buildTime.to = canceledAt;
      break;
  }

  return buildTime;
}

export function buildStatus(build) {
  const { state, createdAt, canceledAt, finishedAt } = build;

  if (state === 'failed' || state === 'passed') {
    return {
      prefix: 'Finished',
      timeValue: finishedAt
    };
  } else if (state === 'blocked') {
    return {
      prefix: 'Blocked',
      timeValue: finishedAt
    };
  } else if (state === 'canceled' || state === 'canceling') {
    return {
      prefix: 'Canceled',
      timeValue: canceledAt
    };
  } else if (state === 'skipped') {
    return {
      prefix: 'Skipped',
      timeValue: createdAt
    };
  }

  return {
    prefix: 'Created',
    timeValue: createdAt
  };
}
