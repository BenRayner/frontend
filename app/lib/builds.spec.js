/* global describe, it, expect */
import { shortMessage, shortCommit, buildTime } from './builds';

describe('shortMessage', () => {
  const BIG_MESSAGE = `âœ¨ Somehow this is the first spec I've written for our front-end, and this seems really odd to me

- this should be a big enough message to make sure we're not running into anything silly`;

  it('returns the first line of a multi-line message', () => {
    expect(shortMessage(BIG_MESSAGE)).toMatchSnapshot();
  });

  it('returns the whole message in a one-line message', () => {
    expect(shortMessage('Bump frontend')).toMatchSnapshot();
  });
});

describe('shortCommit', () => {
  it('returns a 7-character SHA1 for a full SHA1 commitish', () => {
    expect(shortCommit('be9c03b0dbddafccb0882f62182b84924579c93b')).toMatchSnapshot();
  });

  it('returns the full name for a non-SHA1 commitish', () => {
    expect(shortCommit('preload-pipeline-names-sperlunking')).toMatchSnapshot();
  });

  it('returns the full name for a 40-character non-SHA1 commitish', () => {
    expect(shortCommit('so-this-is-a-forty-character-branch-name')).toMatchSnapshot();
  });
});

describe('buildTime', () => {
  it('returns correct time properties for the `started` state', () => {
    expect(buildTime({
      state: 'started',
      startedAt: new Date(1475643466920)
    })).toMatchSnapshot();
  });

  it('returns correct time properties for the `failed` state', () => {
    expect(buildTime({
      state: 'failed',
      startedAt: new Date(1475643466920),
      finishedAt: new Date(1475643467000)
    })).toMatchSnapshot();
  });

  it('returns correct time properties for the `passed` state', () => {
    expect(buildTime({
      state: 'passed',
      startedAt: new Date(1475643466920),
      finishedAt: new Date(1475643467000)
    })).toMatchSnapshot();
  });

  it('returns correct time properties for the `canceled` state', () => {
    expect(buildTime({
      state: 'canceled',
      startedAt: new Date(1475643466920),
      canceledAt: new Date(1475643467000)
    })).toMatchSnapshot();
    expect(buildTime({ state: 'canceled' })).toMatchSnapshot();
  });

  it('returns correct time properties for the `canceling` state', () => {
    expect(buildTime({
      state: 'canceling',
      startedAt: new Date(1475643466920)
    })).toMatchSnapshot();
  });

  it('returns correct time properties for the `scheduled` state', () => {
    expect(buildTime({
      state: 'scheduled',
      scheduledAt: new Date(1475644466920)
    })).toMatchSnapshot();
  });

  it('returns correct time properties for the `skipped` state', () => {
    expect(buildTime({ state: 'skipped' })).toMatchSnapshot();
  });

  it('returns correct time properties for the `not_run` state', () => {
    expect(buildTime({ state: 'not_run' })).toMatchSnapshot();
  });
});

