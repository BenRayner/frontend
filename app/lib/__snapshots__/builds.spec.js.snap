exports[`buildStatus returns correct build status for the \`blocked\` state 1`] = `
Object {
  "prefix": "Blocked",
  "timeValue": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildStatus returns correct build status for the \`canceled\` state 1`] = `
Object {
  "prefix": "Canceled",
  "timeValue": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildStatus returns correct build status for the \`canceling\` state 1`] = `
Object {
  "prefix": "Canceled",
  "timeValue": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildStatus returns correct build status for the \`failed\` state 1`] = `
Object {
  "prefix": "Finished",
  "timeValue": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildStatus returns correct build status for the \`not_run\` state 1`] = `
Object {
  "prefix": "Created",
  "timeValue": 2016-10-05T04:57:46.920Z,
}
`;

exports[`buildStatus returns correct build status for the \`passed\` state 1`] = `
Object {
  "prefix": "Finished",
  "timeValue": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildStatus returns correct build status for the \`scheduled\` state 1`] = `
Object {
  "prefix": "Created",
  "timeValue": 2016-10-05T04:57:46.920Z,
}
`;

exports[`buildStatus returns correct build status for the \`skipped\` state 1`] = `
Object {
  "prefix": "Skipped",
  "timeValue": 2016-10-05T04:57:46.920Z,
}
`;

exports[`buildStatus returns correct build status for the \`started\` state 1`] = `
Object {
  "prefix": "Created",
  "timeValue": 2016-10-05T04:57:46.920Z,
}
`;

exports[`buildTime returns correct time properties for the \`blocked\` state 1`] = `Object {}`;

exports[`buildTime returns correct time properties for the \`canceled\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildTime returns correct time properties for the \`canceled\` state 2`] = `Object {}`;

exports[`buildTime returns correct time properties for the \`canceling\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`buildTime returns correct time properties for the \`failed\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildTime returns correct time properties for the \`not_run\` state 1`] = `Object {}`;

exports[`buildTime returns correct time properties for the \`passed\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`buildTime returns correct time properties for the \`scheduled\` state 1`] = `
Object {
  "from": 2016-10-05T05:14:26.920Z,
}
`;

exports[`buildTime returns correct time properties for the \`skipped\` state 1`] = `Object {}`;

exports[`buildTime returns correct time properties for the \`started\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`shortCommit returns a 7-character SHA1 for a full SHA1 commitish 1`] = `"be9c03b"`;

exports[`shortCommit returns the full name for a 40-character non-SHA1 commitish 1`] = `"so-this-is-a-forty-character-branch-name"`;

exports[`shortCommit returns the full name for a non-SHA1 commitish 1`] = `"preload-pipeline-names-sperlunking"`;

exports[`shortMessage returns the first line of a multi-line message 1`] = `"âœ¨ Somehow this is the first spec I\'ve written for our front-end, and this seems really odd to me"`;

exports[`shortMessage returns the whole message in a one-line message 1`] = `"Bump frontend"`;
